### PR TITLE
MAINT: pyproject.toml, NEP29, Github Actions updates

### PR DESCRIPTION
**Pull request by jklenzing**
# Description

Addresses https://github.com/pysat/pysat/issues/1051

Support for setup.py in pip has been deprecated.  This brings the package up to standard.
- Moves metadata and settings to pyproject.toml
- flake8 settings in setup.cfg
- Uses optional install syntax to deploy test and doc suites to github actions
- updates GA for new versions
- Fixes issue where NEP29 case reinstalls latest numpy anyway
- Update installation instructions

Changes in development flow
- version now specified in pyproject.toml. Code extracts `__version__` from this
- requirements files no longer needed.  maintained for python 3.6 systems
- local installation through `pip install .` or `pip install -e .` for development case
- build versions for pypi via `python -m build`
- pytest coverage options stored in pyproject.toml, will run by default

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

pip install -i https://test.pypi.org/simple/ pysatCDAAC==0.0.3.0

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes
- [x] Update zenodo.json file for new code contributors

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -12,32 +12,34 @@ jobs:
...
+        pip install .[test]
 
-    - name: Install NEP29 dependencies
+    - name: Install with NEP29 dependencies
       if: ${{ matrix.numpy_ver != 'latest'}}
       run: |
         pip install --no-binary :numpy: numpy==${{ matrix.numpy_ver }}
+        # Pandas 1.5 does not support NEP29
+        pip install "pandas<1.5"
+        pip install  --upgrade-strategy only-if-needed .[test]
```
Comment (jklenzing) [inline .github/workflows/main.yml:42]: Need to force install any packages that do not support the minimum numpy.  The upgrade strategy will keep pip from updating numpy.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[flake8]
+max-line-length = 80
+ignore =
+  D200
+  D202
+  W503
```
Comment (jklenzing) [inline tox.ini:6]: flake8 is the one outlier that can't be configured via pyproject.toml.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -12,32 +12,34 @@ jobs:
...
+    - uses: actions/checkout@v3
     - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v2
+      uses: actions/setup-python@v4
       with:
         python-version: ${{ matrix.python-version }}
 
-    - name: Install standard dependencies
+    - name: Install with standard dependencies
+      if: ${{ matrix.numpy_ver == 'latest'}}
```
Comment (jklenzing) [inline .github/workflows/main.yml:32]: Note we skip this now if NEP29 is in play

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,80 @@
+[build-system]
+requires = ["setuptools", "pip >= 10"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "pysatCDAAC"
+version = "0.0.3.0"
```
Comment (jklenzing) [inline pyproject.toml:7]:
```suggestion
version = "0.0.3"
```
The extra digit was to create a unique identifier for the test pypi

**Comment by jklenzing** [comment]
Putting this up for discussion.  We should probably test this out on other packages before merging to be sure we have a consistent way forward.

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+description = "pysat support for CDAAC Instruments"
+readme = "README.md"
+requires-python = ">=3.8"
+license = {file = "LICENSE"}
+authors = [
+    {name = "Jeff Klenzing, et al", email = "pysat.developers@gmail.com"},
+]
```
Comment (jklenzing) [inline pyproject.toml:14]: We also have the option of specifying multiple authors now!

File: pysatCDAAC/__init__.py [hunk 6]
```diff
@@ -1 +1,12 @@
...
+This is a library of `pysat` instrument modules and methods designed to support
+COSMIC instruments and missions archived at the CDAAC web portal.
+
+"""
+
+import pkg_resources
+
 from pysatCDAAC import instruments  # noqa F401
+
+__version__ = pkg_resources.get_distribution('pysatCDAAC').version
```
Comment (jklenzing) [inline pysatCDAAC/__init__.py:12]: `pkg_resources` is core and extracts info from pyproject.toml

**Review by rstoneback — COMMENTED** [review]
Looks fine to me. The pull is still marked as draft. Any further planned changes?

**Comment by jklenzing** [comment]
> Looks fine to me. The pull is still marked as draft. Any further planned changes?

Keeping it as draft until our next meeting. Figured we should all chat before major changes.

**Comment by jklenzing** [comment]
@aburrell, @rstoneback , @JonathonMSmith this is ready for review.  Updated based on the changes in pysatNASA.  Note that we don't have docs yet in this code (see #37).

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -12,11 +12,11 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-latest", "windows-latest"]
-        python-version: ["3.9", "3.10"]
+        python-version: ["3.9", "3.10", "3.11"]
         numpy_ver: ["latest"]
         include:
```
Comment (aburrell) [inline .github/workflows/main.yml:17]: The operations level test should be added.

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,81 @@
+[build-system]
+requires = ["setuptools", "pip >= 10"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "pysatCDAAC"
+version = "0.0.3"
+description = "pysat support for CDAAC Instruments"
+readme = "README.md"
+requires-python = ">=3.8"
```
Comment (aburrell) [inline pyproject.toml:10]:
```suggestion
requires-python = ">=3.6"
```

File: pysatCDAAC/__init__.py [hunk 9]
```diff
@@ -1 +1,18 @@
...
+COSMIC instruments and missions archived at the CDAAC web portal.
+
+"""
+
+import importlib
+import importlib_metadata
+
 from pysatCDAAC import instruments  # noqa F401
+
+# set version
```
Comment (aburrell) [inline pysatCDAAC/__init__.py:13]:
```suggestion
# Set version
```

**Review by aburrell — APPROVED** [review]

**Pull request merged**